boot/env.ml

```diff
@@ -0,0 +1,83 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+module Path = struct
+
+  type t = string
+
+  let relative = Filename.concat
+  let to_string x = x
+  let exists x = Sys.file_exists x
+
+end
+
+(* For now just a pointer to coq/lib (for .vo) and coq-core/lib (for plugins)  *)
+type t =
+  { core: Path.t
+  ; lib : Path.t
+  }
+
+(* fatal error *)
+let fail_msg =
+  "cannot guess a path for Coq libraries; please use -coqlib option \
+   or ensure you have installed the package contaning Coq's stdlib (coq-stdlib in OPAM) \
+   If you intend to use Coq without a standard library, the -boot -noinit options must be used."
+
+let fail s = Format.eprintf "%s@\n%!" fail_msg; exit 1
+
+(* This code needs to be refactored, for now it is just what used to be in envvars  *)
+let guess_coqlib () =
+  Util.getenv_else "COQLIB" (fun () ->
+  let prelude = "theories/Init/Prelude.vo" in
+  Util.check_file_else
+    ~dir:Coq_config.coqlibsuffix
+    ~file:prelude
+    (fun () ->
+      if Sys.file_exists (Filename.concat Coq_config.coqlib prelude)
+      then Coq_config.coqlib
+      else fail ()))
+
+(* Should we fail on double initialization? That seems a way to avoid
+   mis-use for example when we pass command line arguments *)
+let init () =
+  let lib = guess_coqlib () in
+  let core = Util.getenv_else "COQCORELIB"
+      (fun () -> Path.relative lib "../coq-core") in
+  { core ; lib }
+
+let env_ref = ref None
+
+let init () =
+  match !env_ref with
+  | None ->
+    let env = init () in
+    env_ref := Some env; env
+  | Some env -> env
+
+let set_coqlib lib =
+  let env = { lib; core = Path.relative lib "../coq-core" } in
```

Comment:
IIUC this means if we pass -coqlib then COQCORELIB is ignored, is that really what we want?

Reply:
Indeed, rules here are still a bit messy; the plan was to move the handling of `-coqlib` to this lib too, but meanwhile I'll update the PR so if `COQCORELIB` is set, we use it; but it could be the case that we are using `-coqlib` to set both I think.

Reply:
I have changed opinion and I'll try to cleanup this in upcoming PRs, I will properly document what is happening tho.